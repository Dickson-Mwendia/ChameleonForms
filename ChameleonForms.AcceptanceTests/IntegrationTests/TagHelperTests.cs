using System.IO;
using System.Net.Http;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using ApprovalTests.Html;
using ApprovalTests.Reporters;
using ChameleonForms.AcceptanceTests.Helpers;
using Microsoft.AspNetCore.Mvc.Testing;
using Xunit;
using Xunit.Abstractions;

namespace ChameleonForms.AcceptanceTests.IntegrationTests
{
    [UseReporter(typeof(DiffReporter))]
    public class TagHelperTests : IClassFixture<WebApplicationFactory<Example.Startup>>
    {
        private readonly HttpClient _client;
        private readonly WebApplicationFactory<Example.Startup>
            _factory;

        private readonly ITestOutputHelper _outputHelper;

        public TagHelperTests(WebApplicationFactory<Example.Startup> factory)
        {
            _factory = factory;
            _client = factory.CreateClient(new WebApplicationFactoryClientOptions
            {
                AllowAutoRedirect = false
            });
        }

        [Fact]
        public async Task Should_render_correctly()
        {
            var renderedSource = await GetRenderedSourceAsync("/Comparison/ChameleonFormsTH");
            HtmlApprovals.VerifyHtml($"ChameleonFormsTH.cshtml\r\n\r\n{GetViewContents("Comparison/ChameleonFormsTH")}\r\n=====\r\n\r\nRendered Source\r\n\r\n{renderedSource}");
        }

        private async Task<string> GetRenderedSourceAsync(string url)
        {
            var defaultPage = await _client.GetAsync(url);
            var content = await HtmlHelpers.GetDocumentAsync(_client, defaultPage);

            var renderedSource = content.Body.InnerHtml;
            var getFormContent = new Regex(@".*?(<form(.|\n|\r)+?<\/form>).*", RegexOptions.Multiline | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
            return getFormContent.Match(renderedSource).Groups[1].Value;
        }

        private string GetViewContents(string viewPath)
        {
            return File.ReadAllText(string.Format(ViewPath, viewPath));
        }

        private static readonly string ViewPath = Path.Combine(Path.GetDirectoryName(typeof(PartialForTests).Assembly.CodeBase.Replace("file:///", "")), "..", "..", "..", "..", "ChameleonForms.Example", "Views", "{0}.cshtml");
    }
}
