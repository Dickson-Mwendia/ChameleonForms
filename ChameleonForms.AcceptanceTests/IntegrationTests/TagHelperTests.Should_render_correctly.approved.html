ChameleonFormsTH.cshtml

@model ComparisonController.SignupViewModel

@{
    ViewBag.Title = "ChameleonForms Example";
}

<h1>ChameleonForms TagHelpers example</h1>

<chameleon-form attr-id="signup-form">
    <form-message type="Information" heading="Signup for an account">
        <message-paragraph>Please fill in your information below to signup for an account.</message-paragraph>
    </form-message>

    <form-section heading="Your details">
        <field for="FirstName" />
        <field for="LastName" />
        <field for="DateOfBirth" hint="DD/MM/YYYY" />
    </form-section>

    <form-section heading="Account details">
        <field for="EmailAddress" hint="An email will be sent to this address to confirm you own it" />
        <field for="Password" />
        <field for="MembershipType" />
    </form-section>

    <form-section heading="Additional details">
        <field for="Bio" />
        <field for="Homepage" placeholder="http://" />
    </form-section>

    <form-message type="Action" heading="Confirm the Terms & Conditions">
        <message-paragraph>Please <a href="#">read the terms and conditions</a></message-paragraph>
        <field-element for="TermsAndConditions" inline-label="I agree to the terms and conditions" />
    </form-message>

    <form-navigation>
        <submit-button label="Signup" emphasis-style="Primary" />
    </form-navigation>
</chameleon-form>

=====

Rendered Source

<form action="" enctype="application/x-www-form-urlencoded" id="signup-form" method="post" novalidate="novalidate">
          <div class="information_message">
          <h3>Signup for an account</h3>
          <div class="message">

        
<p>
    Please fill in your information below to signup for an account.
</p>
              </div>
      </div>


        <fieldset>
        <legend>Your details</legend>
        <dl>

                    <dt><label for="FirstName">First name</label> <em class="required">*</em></dt>
            <dd>
                <input data-val="true" data-val-required="The First name field is required." id="FirstName" name="FirstName" required="required" type="text" value=""> <span class="field-validation-valid" data-valmsg-for="FirstName" data-valmsg-replace="true"></span>
            </dd>

                    <dt><label for="LastName">Last name</label> <em class="required">*</em></dt>
            <dd>
                <input data-val="true" data-val-required="The Last name field is required." id="LastName" name="LastName" required="required" type="text" value=""> <span class="field-validation-valid" data-valmsg-for="LastName" data-valmsg-replace="true"></span>
            </dd>

                    <dt><label for="DateOfBirth">Date of birth</label> <em class="required">*</em></dt>
            <dd>
                <input aria-describedby="DateOfBirth--Hint" data-val="true" data-val-date="The field Date of birth must be a date with format d/M/yyyy." data-val-format="d/M/yyyy" data-val-required="The Date of birth field is required." id="DateOfBirth" name="DateOfBirth" required="required" type="text" value=""><div class="hint" id="DateOfBirth--Hint">DD/MM/YYYY</div> <span class="field-validation-valid" data-valmsg-for="DateOfBirth" data-valmsg-replace="true"></span>
            </dd>

            </dl>
    </fieldset>


        <fieldset>
        <legend>Account details</legend>
        <dl>

                    <dt><label for="EmailAddress">Email address</label> <em class="required">*</em></dt>
            <dd>
                <input aria-describedby="EmailAddress--Hint" data-val="true" data-val-email="The Email address field is not a valid e-mail address." data-val-required="The Email address field is required." id="EmailAddress" name="EmailAddress" required="required" type="email" value=""><div class="hint" id="EmailAddress--Hint">An email will be sent to this address to confirm you own it</div> <span class="field-validation-valid" data-valmsg-for="EmailAddress" data-valmsg-replace="true"></span>
            </dd>

                    <dt><label for="Password">Password</label> <em class="required">*</em></dt>
            <dd>
                <input data-val="true" data-val-required="The Password field is required." id="Password" name="Password" required="required" type="password"> <span class="field-validation-valid" data-valmsg-for="Password" data-valmsg-replace="true"></span>
            </dd>

                    <dt><label for="MembershipType">Membership type</label> <em class="required">*</em></dt>
            <dd>
                <select data-val="true" data-val-required="The Membership type field is required." id="MembershipType" name="MembershipType" required="required"><option selected="selected" value="Standard">Standard</option>
<option value="Bonze">Bonze</option>
<option value="Silver">Silver</option>
<option value="Gold">Gold</option>
<option value="Platinum">Platinum</option>
</select> <span class="field-validation-valid" data-valmsg-for="MembershipType" data-valmsg-replace="true"></span>
            </dd>

            </dl>
    </fieldset>


        <fieldset>
        <legend>Additional details</legend>
        <dl>

                    <dt><label for="Bio">Bio</label></dt>
            <dd>
                <textarea cols="20" id="Bio" name="Bio" rows="2"></textarea> <span class="field-validation-valid" data-valmsg-for="Bio" data-valmsg-replace="true"></span>
            </dd>

                    <dt><label for="Homepage">Homepage</label></dt>
            <dd>
                <input id="Homepage" name="Homepage" placeholder="http://" type="url" value=""> <span class="field-validation-valid" data-valmsg-for="Homepage" data-valmsg-replace="true"></span>
            </dd>

            </dl>
    </fieldset>


          <div class="action_message">
          <h3>Confirm the Terms &amp; Conditions</h3>
          <div class="message">

        
<p>
    Please <a href="#">read the terms and conditions</a>
</p>
        <input data-val="true" data-val-required="The Terms and conditions field is required." id="TermsAndConditions" name="TermsAndConditions" required="required" type="checkbox" value="true"> <label for="TermsAndConditions">I agree to the terms and conditions</label>
              </div>
      </div>


            <div class="form_navigation">

        <button class="btn-primary" type="submit">Signup</button>
            </div>

    </form>